### PR TITLE
TypeScript Migration

### PR DESCRIPTION
**Pull request by seijiwjs**

### CODE REVIEW

File: src/Handles/AxialRotationHandle.ts [hunk 1]
```diff
@@ -10,14 +23,18 @@ import './Shaders/HandleShader'
  * const xfo1 = new Xfo()
  * // This is rotation over `Y` axis
  * xfo1.ori.setFromAxisAndAngle(new Vec3(0, 1, 0), Math.PI * 0.5)
- * axialRotationHandle.getParameter('LocalXfo').setValue(xfo1)
+ * axialRotationHandle.localXfoParam.value = (xfo1)
```
Comment (phtaylor) [inline src/Handles/AxialRotationHandle.ts:26]: (xfo) ??

File: src/Handles/BaseAxialRotationHandle.ts [hunk 2]
```diff
@@ -40,25 +57,24 @@ class BaseAxialRotationHandle extends Handle {
    *
    * @return {Parameter} - returns handle's target global Xfo.
    */
-  getTargetParam() {
-    return this.param ? this.param : this.getParameter('GlobalXfo')
+  getTargetParam(): Parameter<unknown> {
+    return this.param ? this.param : this.globalXfoParam
   }
```
Comment (phtaylor) [inline src/Handles/BaseAxialRotationHandle.ts:62]: NumberParameter | XfoParameter

File: src/Handles/LinearMovementHandle.ts [hunk 3]
```diff
@@ -86,18 +105,18 @@ class LinearMovementHandle extends BaseLinearMovementHandle {
...
 
   /**
    * Handles the initially drag of the handle.
    *
    * @param {MouseEvent|TouchEvent|object} event - The event param.
    */
-  onDragStart(event) {
-    this.grabPos = event.grabPos
+  onDragStart(event: ZeaPointerEvent) {
+    this.grabPos = this.grabPos
```
Comment (phtaylor) [inline src/Handles/LinearMovementHandle.ts:117]: this.grabPos = this.grabPos ??

File: src/Measurement/MeasureAngle.ts [hunk 4]
```diff
@@ -51,8 +62,8 @@ class MeasureAngle extends TreeItem {
     this.markerMaterialB.getParameter('Overlay').setValue(0.5)
 
     this.lineMaterial = new LinesMaterial('Line')
-    this.lineMaterial.getParameter('BaseColor').setValue(new Color(0, 0, 0))
-    this.lineMaterial.getParameter('Overlay').setValue(0.5)
+    this.lineMaterial.baseColorParam.setValue(new Color(0, 0, 0))
+    this.lineMaterial.overlayParam.setValue(0.5)
```
Comment (phtaylor) [inline src/Measurement/MeasureAngle.ts:67]: maybe .value = new Color(0, 0, 0)?

File: src/Measurement/MeasureAngle.ts [hunk 5]
```diff
@@ -115,46 +126,46 @@ class MeasureAngle extends TreeItem {
     labelXfo.tr.scaleInPlace(0.5)
 
     xfoA.ori.setFromDirectionAndUpvector(tangentA, normA)
-    this.markerA.getParameter('GlobalXfo').setValue(xfoA)
+    this.markerA.globalXfoParam.setValue(xfoA)
     xfoB.ori.setFromDirectionAndUpvector(tangentB, normA)
-    this.markerB.getParameter('GlobalXfo').setValue(xfoB)
+    this.markerB.globalXfoParam.setValue(xfoB)
```
Comment (phtaylor) [inline src/Measurement/MeasureAngle.ts:132]: value = xfoA

File: src/Measurement/MeasureAngle.ts [hunk 6]
```diff
@@ -115,46 +126,46 @@ class MeasureAngle extends TreeItem {
...
-    this.markerA.getChild(0).getParameter('LocalXfo').setValue(lineAXfo)
+    this.markerA.getChild(0).localXfoParam.value = lineAXfo
     const lineBXfo = new Xfo()
     lineBXfo.sc.z = params[1]
-    this.markerB.getChild(0).getParameter('LocalXfo').setValue(lineBXfo)
+    this.markerB.getChild(0).localXfoParam.value = lineBXfo
 
-    this.billboard.getParameter('GlobalXfo').setValue(labelXfo)
+    this.billboard.globalXfoParam.setValue(labelXfo)
   }
```
Comment (phtaylor) [inline src/Measurement/MeasureAngle.ts:141]: here too? Maybe eliminate all the .setValue() calls that are left.

File: src/SelectionGroup.ts [hunk 7]
```diff
@@ -22,14 +22,16 @@ const GROUP_XFO_MODES = {
  * @extends {SelectionSet}
  */
 class SelectionGroup extends SelectionSet {
+  initialXfoModeParam
+  selectionGroupXfoOp
```
Comment (phtaylor) [inline src/SelectionGroup.ts:26]: types?

Comment (seijiwjs) [reply src/SelectionGroup.ts:26]: Ok, just fixed this

File: src/UndoRedo/Changes/TreeItemAddChange.ts [hunk 8]
```diff
@@ -92,25 +97,23 @@ class TreeItemAddChange extends Change {
...
       return
     }
     this.name = j.name
     this.treeItem = treeItem
-    this.treeItem.addRef(this)
 
     this.treeItem.fromJSON(j.treeItem, context)
-    this.treeItemIndex = this.owner.addChild(this.treeItem, false, false)
+    this.treeItemIndex = this.owner.getChildIndex(this.owner.addChild(this.treeItem))
   }
```
Comment (phtaylor) [inline src/UndoRedo/Changes/TreeItemAddChange.ts:111]: curious about this line. I didn't realize 'addChild' returned the child , and not the index.

**Review by phtaylor — APPROVED** [review]

File: src/UndoRedo/Changes/TreeItemAddChange.ts (hunk 8 above)
Comment (seijiwjs) [reply src/UndoRedo/Changes/TreeItemAddChange.ts:111]: Yeah, I found this in a few areas.

**Pull request merged**